File: team.py

from tqdm import tqdm

PART_SIZE=100

def solve2(teams, pizzas):
    if len(pizzas) < 1:
        return
    for team in tqdm(teams):
        if len(pizzas) < 1:
            break
        maxScore = 0
        maxScoreIdx = 0
        for i, pizza in enumerate(pizzas[0:PART_SIZE]):
            score = team.calcSc(pizza)
            if score > maxScore:
                maxScore = score
                maxScoreIdx = i
        team.add(pizzas[maxScoreIdx])
        pizzas.pop(maxScoreIdx)

def solve(teams, pizzas):
    if len(pizzas) < 1:
        return
    for team in tqdm(teams):
        if len(pizzas) < 1:
            break
        for i in range(team.cap):
            maxScore = 0
            maxScoreIdx = 0
            for i, pizza in enumerate(pizzas[0:PART_SIZE]):
                score = team.calcSc(pizza)
                if score > maxScore:
                    maxScore = score
                    maxScoreIdx = i
            team.add(pizzas[maxScoreIdx])
            pizzas.pop(maxScoreIdx)




class Team():
    def __init__(self, cap):
        self.cap = cap
        self.pizzas = []
        self.ings = set()

    # @staticmethod
    # def calcScore(pizzas):
    #     uniq = set.union( [p.ings for p in pizzas] )
    #     total = 0
    #     for p in pizzas:
    #         total += p.count
    #     return len(uniq) / total

    def calcSc(self, pizza):
        comm = self.ings.intersection(pizza.ings)
        uniq = self.ings.union(pizza.ings)
        total = pizza.count
        for p in self.pizzas:
            total += p.count
        sc = len(uniq) - len(comm) #(len(uniq)**2)/ (total)
        return sc

    def add(self, pizza):
        assert 1 + len(self.pizzas) <= self.cap
        self.pizzas.append(pizza)
        assert pizza.selected == False
        pizza.selected = True
        self.ings = self.ings.union(pizza.ings)
    
    def __repr__(self):
        return '[{}-{}-{}]'.format(self.cap, [p.index for p in self.pizzas], self.ings)

    @property
    def is_full(self):
        return len(self.pizzas) == self.cap

File: go/team.go
package main

import "fmt"

// Team is
type Team struct {
	index  int
	cap    int
	pizzas []*Pizza
	ings   map[string]bool
}

// NewTeam is
func NewTeam(index int, cap int) Team {
	t := Team{}
	t.index = index
	t.cap = cap
	t.pizzas = []*Pizza{}
	t.ings = map[string]bool{}
	return t
}

func addPizza(team *Team, pizza *Pizza) {
	if pizza.selected {
		panic("Pizza already selected.")
	}
	if len(team.pizzas)+1 > team.cap {
		panic("Team is full")
	}
	team.pizzas = append(team.pizzas, pizza)
	for k := range pizza.ings {
		team.ings[k] = true
	}
	pizza.selected = true
	fmt.Println("Pizza:", pizza.index, ":sc", pizza.count, "added to Team:", team.index)
}

func calcSc(team *Team, pizza *Pizza) int {
	commCount := 0
	for k := range pizza.ings {
		_, ok := team.ings[k]
		if ok {
			commCount++
		}
	}
	return len(pizza.ings) + len(team.ings) - commCount
}
